connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,265 @@
+goal of the improvement is to reduce the number of RTTs that takes to select the
+muxer of a connection. The solution relies on the ability of the security
+protocol's handshake process to negotiate higher level protocols, which enables
+the muxer selection to be carried out along with security protocol handshake.
+The proposed solution saves the RTT of multistream selection for muxers.
+
+For more context and the status of this work, please refer to [#426]
+
+
+## Design
+
+### Current connection upgrade process
+
+The current connection upgrade process is described in detail in [connections].
+As shown in this [sequence-chart], after a network connection is established,
+the following will happen to upgrade the conntection to a capable connection.
+
+1. The multistream-selection protocol is run over the connection to select the
+security protocol to be used.
+2. The selected security protocol performs handshaking and establishs a secure
+tunnel
+3. The multistream-selection protocol then will run again for muxer selection.
+4. Connection then is upgraded to a capable connection by the selected muxer.
+
+
+## Improved muxer selection
+
+The security protocol's ability of supporting higher level abstract protocol
+negotiation (for example, TLS's support of ALPN, and Noise's support of Early
+Data) makes it possible to collapse the step 2 and step 3 in the previous
+section into one step. Muxer selection can be performed as part of the security
+protocol handshake, thus there is no need to perform another mutistream
+-selection negotiation for muxer selection.
+
+In order to achieve the above stated goal, each candidate muxer will be
+represented by a protocol name/code, and the candidate muxers are supplied to
+the security protocol's handshake process as a list of protocol names.
+
+If the client and server agree upon the common muxer to be used, then the
+result of the muxer selection is a muxer code represented by the selected
+protocol name. If no agreement is reached upon by the client and server
+then an empty muxer code is returned and the connection upgrade process
+MUST fall back to the multistream-selection protocol to negotiate the muxer.
+
+
+### Muxer selection over TLS
+
+When the security protocol selected by the upgrader is TLS, the [ALPN]
+extesion of TLS handshake is used to select the muxer.
+
+   With ALPN, the client sends the list of supported application
+   protocols as part of the TLS ClientHello message.  The server chooses
+   a protocol and sends the selected protocol as part of the TLS
+   ServerHello message.  The application protocol negotiation can thus
+   be accomplished within the TLS handshake, without adding network
+   round-trips, and allows the server to associate a different
+   certificate with each application protocol, if desired.
+
+For the purpose of muxer selection, the types of muxers are coded as protocol
+names in the form of a list of strings, and inserted in the ALPN "NextProtos"
```

Comment:
“NextProtos” is the name of the field in the Go implementation (for historic reasons, because it was first an implementation of NPN, not ALPN). I suggest sticking with the RFC terminology here.

Reply:
good point, changed that to "ALPN extesion"

Reply:
thanks for the catch, agreed. changed to "ALPN extension" which is aligned with the RFC